### DISCUSSION TITLE
How can I reset the provider if a user signed out

### DISCUSSION THREAD

**Discussion by desmeit** [rrousselGit/provider#689] [Q&A]
If a user signed out and a different user logs in, the value of the streamprovider is not changing.

My structure:

main widget:

```
 @override
  Widget build(BuildContext context) {
    return MultiProvider(
        providers: [
          Provider<AuthenticationService>(
            create: (_) => AuthenticationService(FirebaseAuth.instance),
          ),
          StreamProvider(
            create: (context) =>
                context.read<AuthenticationService>().authStateChanges,
            initialData: null,
          ),
          StreamProvider<UserData>(
            create: (context) => FirebaseService(
                    uid: Provider.of<AuthenticationService>(context,
                            listen: false)
                        .getCurrentUID())
                .userData,
            initialData: null,
            // catchError: (_, __) => null,
          ),
         ],
         child: MaterialApp(
         initialRoute: '/',
         onGenerateRoute: RouteGenerator.generateRoute,
...
```

the route generator:

```
 switch (settings.name) {
      case '/':
        return MaterialPageRoute(
            builder: (_) => AuthenticationWrapper(data: args));
```

the AuthenticationWrapper:

```
  @override
  Widget build(BuildContext context) {
    print("AuthenticationWrapper rebuilds");
    final firebaseUser = context.watch<User>();

 if (firebaseUser != null) {
      print('uid: ' + firebaseUser.uid);

      return CupertinoHomePage(data: widget.data);
...
```

This is working but at some point I want to sign out the user:

```
context.read<AuthenticationService>().signOut();
Navigator.pop(context);
```
After the logout I want to sign in a different user but the userData Stream is still the old one. I think it has to do with the fact, that the provider ist in the main Widget above the MaterialApp and also above the / route (AuthenticationWrapper). How can I reset the Provider? If I put it in the AuthenticationWrapper then it could work I think but then I have the problem that the provider is not working anymore across all routes.

What I'm doing wrong?

**Comment by rrousselGit** [comment]
Your problem is the `create` of your Streamprovider: 

```dart
         create: (context) => FirebaseService(
                    uid: Provider.of<AuthenticationService>(context,
                            listen: false)
                        .getCurrentUID())
```


Since you're passing the UUID as parameter to FirebaseService, you're never dealing with cases where the UUID changes

You should instead create a Stream that not only listens to the user data but also changes to the user ID.

**Reply by desmeit** [reply]
@rrousselGit 
I'm trying to test your proposed solution again: Where exactly do you think the stream should be? In the provider or in the UserData stream? Maybe you can give a minimal example?

**Comment by desmeit** [comment]
I solved it this way:

```
Navigator.pushAndRemoveUntil(
                      context,
                      MaterialPageRoute(
                          builder: (BuildContext context) => MyApp()),
                      (Route<dynamic> route) => false);
```

Instead of pushing to the / route I push it now to the main widget. This rebuilds the whole app.
I'm not sure if this is the best solution but it works. What do you think?
